### PR TITLE
Wma/fix 488

### LINKED ISSUE #488: should we support fsprand(n, m, p) when n * m > typemax(Int)?
Hi @willow-ahrens,

I'm working on benchmarks and for `1_000_000x1_000_000` I'm not able to move from COO format (that is returned by `fsprand`) to `CSC`:

```julia
using Finch

M = Tensor(Dense(SparseList(Element(0.0))), fsprand(1_000_000, 1_000_000, 1e-05));
```
where the last line runs infinitely.

### PR DESCRIPTION
**Pull request by willow-ahrens**
fixes #488

### CODE REVIEW

**Comment by willow-ahrens** [comment]
It looks like fsprand is erroring due to integer overflow on 32bit systems with this test case.

**Comment by willow-ahrens** [comment]
this is due to randsubseq, not Finch. We need to implement our own randsubseq or improve the existing implementation

**Comment by mtsokol** [comment]
It looks that it's strictly due to 32bit integer being a default one on Windows.

It seems that `length(A)` gets an overflow and returns `-727379968`. But then isn't it a bug in Julia?

```julia
length(CartesianIndices((1000000, 1000000)))
```
Gives `-727379968` of 32bit integer. But I think it should anticipate which size to use and return 64bit integer.

**Comment by willow-ahrens** [comment]
the main challenge is that all of these algorithms assume the entire collection we are sampling from is representable in memory.

**Comment by mtsokol** [comment]
I added custom implementation to make sure index variables are `Int64` for now. Weirdly it fails with:
```
BoundsError: attempt to access 1000000×1000000 CartesianIndices{2, Tuple{Base.OneTo{Int32}, Base.OneTo{Int32}}} at index [38056]
```
which doesn't really make any sense, considering that 38056 fits in Int32.

**Comment by mtsokol** [comment]
[I received a feedback on Julia Discourse.](https://discourse.julialang.org/t/boundserror-in-cartesianindices-on-32-bit-windows/112777/4?u=expine) The issue here is that internally it still checks the index against the length, so it looks like it's going to be on 32-bit anyway.

**Comment by mtsokol** [comment]
One option is that we can build index ourselves, in the line where error occurs:
```julia
push!(S, A[i += ceil(Int64, s)])
```
do instead something like:
```julia
i += ceil(Int64, s)
i_tuple = Tuple{N}
for idx, dim in enumerate(dimensions):
    i_tuple[idx] = i % dim
    i = i // dim
push!(S, A[i_tuple...])
```
This way we can index A with a tuple instead of linearized index that can exceed 32bit.

**Comment by willow-ahrens** [comment]
We may want to move towards a custom implementation for tuples of ints

**Comment by willow-ahrens** [comment]
Im also happy to take a crack at this, if you like

**Comment by mtsokol** [comment]
Let me take care of this! It's a small tweak to the existing implementation.

File: src/interface/fsparse.jl [hunk 1]
```diff
@@ -130,6 +130,38 @@ end
...
+
+randsubseq(r::AbstractRNG, A::AbstractArray{T}, p::Real) where {T} =
+    randsubseq!(r, T[], A, p)
+
+# Fill S (resized as needed) with a random subsequence of A, where
+# each element of A is included in S with independent probability p.
+function randsubseq!(r::AbstractRNG, S::AbstractArray, A::AbstractArray, p::Real)
+    Base.require_one_based_indexing(S, A)
+    0 <= p <= 1 || throw(ArgumentError("probability $p not in [0,1]"))
+    n = prod(size(A), init=Int64(1))  # length
```
Comment (willow-ahrens) [inline src/interface/fsparse.jl:143]: This line will overflow (int64 is not enough in the case we want a sparse matrix with int64 indices)

File: src/interface/fsparse.jl [hunk 2]
```diff
@@ -130,6 +130,38 @@ end
...
+    p == 1 && return copyto!(resize!(S, n), A)
```
Comment (willow-ahrens) [inline src/interface/fsparse.jl:144]: this line is fine to use prod(size(A)) because we know it won't overflow because we are about to materialize it.

**Comment by willow-ahrens** [comment]
I'm concerned we need to use a different algorithm. We can't just call `randexp` and convert to an integer, I'm not even sure that's correct for Int64. I don't think we can expect it to perform well after the float mantissa loses precision, around $n=2^53$ this algorithm is only going to select $i$ equal to multiples of 2, 4, 8, etc.

**Comment by mtsokol** [comment]
FYI The original implementation does:
```julia
s = randexp(r) * L
s >= n - i && return S
push!(S, A[i += ceil(Int,s)])
```

**Comment by willow-ahrens** [comment]
Yes, don't think the original algorithm is correct for sparse arrays with large dimension. The problem is that the function `f(x) = ceil(Int, x::Float)` is not actually capable of producing all possible integers.

**Comment by willow-ahrens** [comment]
we might want to use something like https://github.com/JuliaStats/StatsBase.jl/blob/60fb5cd400c31d75efd5cdb7e4edd5088d4b1229/src/sampling.jl#L256C1-L278C44, or we may also want to consider . This would allow us to customize the simple algorithm for tuples of ints, with correct behavior even for `fsprand(typemax(Int64), typemax(Int64), (10.0/typemax(Int64))/typemax(Int64))`. 

The above algorithm is for sampling k numbers without replacement. However, we still need to accurately generate the number of numbers we will need. We need a binomial distribution, but we need to be careful with how we call that too, since we want to sample from way more than n.

However, it's debatable whether we should do any of this, because even though we expect the number of output nonzeros to be addressable, we aren't prepared for the case where the number of elements overflows.

We may want to instead introduce a variant of `fsprand` that allows us to specify the exact number of nonzeros, then use the sampling $k$ without replacement algorithm I linked here.

**Comment by willow-ahrens** [comment]
Here's my issue:

```
julia> using Finch
Precompiling Finch
  1 dependency successfully precompiled in 24 seconds. 13 already precompiled.

julia> t = 1000
1000

julia> countstored(fsprand(t, t, (100/t)/t))
108

julia> countstored(fsprand(t, t, (100/t)/t))
108

julia> countstored(fsprand(t, t, (100/t)/t))
99

julia> t = typemax(Int64)
9223372036854775807

julia> countstored(fsprand(t, t, (100/t)/t))
0

julia> countstored(fsprand(t, t, (100/t)/t))
0

julia> countstored(fsprand(t, t, (100/t)/t))
0
```

**Comment by hameerabbasi** [comment]
We have an implementation here if that helps:
https://github.com/pydata/sparse/blob/0b58d4b023d299750c324f8250ab2cdb661aeea9/sparse/pydata_backend/_utils.py#L223

**Comment by willow-ahrens** [comment]
@hameerabbasi I don't think the algorithm you mentioned can handle this either:

```
willow@Willows-Mac-mini ~ % poetry show 
joblib        1.4.0  Lightweight pipelining with Python functions
llvmlite      0.42.0 lightweight wrapper around basic LLVM functionality
numba         0.59.1 compiling Python code using LLVM
numpy         1.26.4 Fundamental package for array computing in Python
scikit-learn  1.4.2  A set of python modules for machine learning and data mining
scipy         1.13.0 Fundamental algorithms for scientific computing in Python
sparse        0.15.1 Sparse n-dimensional arrays for the PyData ecosystem
threadpoolctl 3.4.0  threadpoolctl
willow@Willows-Mac-mini ~ % poetry run python3
Python 3.11.8 (main, Feb  6 2024, 21:21:21) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sparse
>>> sparse.random([1000000,1000000,1000000,1000000], nnz=4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/willow/Library/Caches/pypoetry/virtualenvs/willow-ENU-A1do-py3.11/lib/python3.11/site-packages/sparse/_utils.py", line 340, in random
    ).reshape(shape)
      ^^^^^^^^^^^^^^
  File "/Users/willow/Library/Caches/pypoetry/virtualenvs/willow-ENU-A1do-py3.11/lib/python3.11/site-packages/sparse/_coo/core.py", line 1029, in reshape
    raise ValueError(f"cannot reshape array of size {self.size} into shape {shape}")
ValueError: cannot reshape array of size 2003764205206896640 into shape (1000000, 1000000, 1000000, 1000000)
```

**Comment by willow-ahrens** [comment]
Just fyi, that number is the overflowed version of 1000000^4

```
julia> t = 1000000
1000000

julia> t * t * t * t
2003764205206896640
```

**Comment by willow-ahrens** [comment]
Does this look good to everyone? There's a lot of statistics here so it would be nice if someone else could review.

**Review by mtsokol — APPROVED** [review]

**Comment by hameerabbasi** [comment]
> Just fyi, that number is the overflowed version of 1000000^4
> 
> 
> 
> ```
> 
> julia> t = 1000000
> 
> 1000000
> 
> 
> 
> julia> t * t * t * t
> 
> 2003764205206896640
> 
> ```

Yes, in `sparse`, we currently don't support arrays of size `>= 2 ** 64`.

**Pull request merged**